protowhat/State.py

```diff
@@ -104,19 +130,32 @@ def to_child(self, append_message="", **kwargs):
 
         return child
 
-    def build_message(self, tail_msg="", fmt_kwargs=None):
-
+    def build_message(self, tail_msg="", fmt_kwargs=None, append=True):
         if not fmt_kwargs:
             fmt_kwargs = {}
         out_list = []
         # add trailing message to msg list
         msgs = self.messages[:] + [{"msg": tail_msg, "kwargs": fmt_kwargs}]
 
-        for d in msgs:
+        # format messages in list, by iterating over previous, current, and next message
+        for prev_d, d, next_d in zip([{}, *msgs[:-1]], msgs, [*msgs[1:], {}]):
```

Comment:
Not blocking, but it seems you're going for a kind of rolling window here. Generally, stuff like this is nice to extract out in a utility. It's easier to test as well. What I'd do is create a rolling window iterator which does exactly this. The for loop here would simplify to:

```python3
for prev_d, d, next_d in rolling_window_it(msgs, 1, {}):
  ...
```

Where `rolling_window_it(msgs, 1, {})` will create a rolling window of 1 on each side `None` is filled in with `{}`.

It will be a bit more verbose, since you'll have to implement the iteror, but is easier to understand and reusable.